ext/cedar_policy/src/request.rs

```diff
@@ -1,32 +1,45 @@
-use cedar_policy::Request;
-use magnus::{function, method, Error, Module, Object, RModule, Ruby, Value};
+use cedar_policy::{Request, Schema};
+use magnus::{
+    function, method,
+    scan_args::{get_kwargs, scan_args},
+    Error, Module, Object, RModule, Ruby, TryConvert, Value,
+};
 use std::convert::Into;
 
 use crate::{
     context::ContextWrapper,
     entity_uid::{to_euid_value, EntityUidWrapper},
+    error::REQUEST_VALIDATION_ERROR,
+    schema::RSchema,
 };
 
 #[magnus::wrap(class = "CedarPolicy::Request")]
 pub struct RRequest(Request);
 
 impl RRequest {
-    fn new(
-        ruby: &Ruby,
-        principal: EntityUidWrapper,
-        action: EntityUidWrapper,
-        resource: EntityUidWrapper,
-        context: ContextWrapper,
-    ) -> Result<Self, Error> {
+    fn new(ruby: &Ruby, args: &[Value]) -> Result<Self, Error> {
+        let args = scan_args::<_, (), (), (), _, ()>(args)?;
+        let (principal, action, resource, context): (
+            EntityUidWrapper,
+            EntityUidWrapper,
+            EntityUidWrapper,
+            ContextWrapper,
+        ) = args.required;
+        let kw_args = get_kwargs::<_, (), (Option<Value>,), ()>(args.keywords, &[], &["schema"])?;
+        let (schema,): (Option<Value>,) = kw_args.optional;
+        let schema = schema
+            .and_then(|s| <&RSchema>::try_convert(s).ok())
+            .map(|s| Schema::from(s));
```

Comment:
Yes, I was able to get the `try_convert` to happen automatically by telling `get_kwargs` that I am expecting an `Option<&RSchema>` instead of an `Option<Value>`. 😄 

In theory, it _seems_ like I should be able to get an `Option<RSchema>` instead of `Option<&RSchema>` and that might possible to avoid a `.clone()` on the Cedar Schema object. However,  it looks like `get_kwargs` can only return types if the implement `TryConvert`... and I don't know how to build a `TryConvert` implementation for `RSchema`.

Reply:
I think `Option<&RSchema>` is fine here. We copy the value to avoid incorrect memory in Rust If you implemented the `TryConvert` trait for `RSchema` I think you can use it instead to call `<&RSchema>::try_convert(s)`